File: Commons/Graph/src/main/java/com/greentree/commons/graph/algorithm/sort/BaseTopologicalSorter.java
package com.greentree.commons.graph.algorithm.sort;

import com.greentree.commons.graph.Graph;
import com.greentree.commons.graph.algorithm.walk.GraphVisitor;
import com.greentree.commons.graph.algorithm.walk.OneUse;

import java.util.Collections;
import java.util.List;

public record BaseTopologicalSorter<V>(Graph<? extends V> graph) implements TopologicalSorter<V> {

    @Override
    public void sort(List<? super V> list) {
        var walker = new OneUse<>(graph);
        walker.visit((GraphVisitor<V>) v -> {
            list.add(0, v);
            return true;
        });
        Collections.reverse(list);
    }

}

File: Commons/Graph/src/main/java/com/greentree/commons/graph/algorithm/sort/BaseTopologicalSorter.kt
package com.greentree.commons.graph.algorithm.sort

import com.greentree.commons.graph.FiniteGraph
import com.greentree.commons.graph.algorithm.pass.DFSGraphPass
import com.greentree.commons.graph.algorithm.pass.asOneUse
import com.greentree.commons.graph.algorithm.walk.vertex.VertexVisitor

data class BaseTopologicalSorter<V : Any>(val graph: FiniteGraph<V>) : TopologicalSorter<V> {

	override fun sort(list: MutableList<in V>) {
		val walker = DFSGraphPass.instance<V>().asOneUse()
		walker.visit(graph, object : VertexVisitor<V> {
			override fun endVisit(v: V) {
				list.add(v)
			}

			override fun startVisit(v: V): Boolean {
				return true
			}
		})
		list.reverse()
	}
}
